sympy/solvers/solveset.py

```diff
@@ -998,6 +1006,203 @@ def _solveset(f, symbol, domain, _check=False):
+
+    Returns
+    =======
+
+    A tuple (f_x, g_n) is being returned where f_x is modular independent function
+    of symbol and g_n being set of values f_x can have.
+
+    """
+    a, m = modterm.args
+
+    if not isinstance(a, (Dummy, Symbol, Add, Mul, Pow)):
+        return modterm, rhs
+
+    if (rhs - m).is_positive or (rhs - m) is S.Zero:
+        return symbol, EmptySet()
+
+    if a is symbol:
+        return symbol, ImageSet(Lambda(n, m*n + rhs), S.Integers)
+
+    if a.is_Add:
+        # g + h = a
+        g, h = a.as_independent(symbol)
+        if g is not S.Zero:
+            return _invert_modular(Mod(h, m), (rhs - Mod(g, m)) % m, n, symbol)
+
+    if a.is_Mul:
+        # g*h = a
+        g, h = a.as_independent(symbol)
+        if g is not S.One:
+            return _invert_modular(Mod(h, m), (rhs*invert(g, m)) % m, n, symbol)
+
+    if a.is_Pow:
+        # base**expo = a
+        base, expo = a.args
+        if expo.has(symbol) and not base.has(symbol):
+            # remainder -> solution independent of n of equation.
+            # m, rhs are made coprime by dividing igcd(m, rhs)
+            remainder = discrete_log(m / igcd(m, rhs), rhs, a.base)
+            # period -> coefficient of n in the solution and also referred as
+            # the least period of expo in which it is repeats itself.
+            # (a**(totient(m)) - 1) divides m. Here is link of theoram:
+            # (https://en.wikipedia.org/wiki/Euler's_theorem)
+            period = totient(m)
+            for p in divisors(period):
+                # there might a lesser period exist than totient(m).
+                if pow(a.base, p, m / igcd(m, a.base)) == 1:
+                    period = p
+                    break
+            return expo, ImageSet(Lambda(n, period*n + remainder), S.Integers)
+
+    return a, ImageSet(Lambda(n, m*n + rhs), S.Integers)
+
+
+def _solve_modular(f, symbol, domain):
+    r"""
+    Helper function for solving modular equations of type ``A - Mod(B, C) = 0``,
+    where A can or cannot be a function of symbol, B is surely a function of
+    symbol and C is an integer.
+
+    Currently ``_solve_modular`` is only able to solve cases
+    where A is not a function of symbol.
+
+    Parameters
+    ==========
+
+    f : Expr
+        The modular equation to be solved, ``f = 0``
+
+    symbol : Symbol
+        The variable in the equation to be solved.
+
+    domain : Set
+        A set over which the equation is solved.
+
+    Returns
+    =======
+
+    A set of solutions satisfying the given modular equation.
+    A ``ConditionSet`` if the equation is unsolvable.
+
+    Examples
+    ========
+
+    >>> from sympy.solvers.solveset import _solve_modular as solve_modulo
+    >>> from sympy import S, Symbol, sin
+    >>> from sympy.core.mod import Mod
+    >>> x = Symbol('x')
+    >>> solve_modulo(Mod(5*x - 8, 7) - 3, x, S.Reals)
+    ImageSet(Lambda(_n, 7*_n + 5), Integers)
+    >>> solve_modulo(-7 + Mod(x, 5), x, S.Complexes)
```

Comment:
What's the meaning of `Mod(x, 5)` when x is complex? It seems to me that this is still open for debate: https://github.com/sympy/sympy/issues/11391
It'd be curious to have a solver for this specific case if `Mod` itself is currently undefined for complex numbers.

Reply:
> What's the meaning of Mod(x, 5) when x is complex?

No meaning it is when x is complex. **But** do note that S.Complexes is just domain I am using in test case, which doesn't meant x to be definetly Complex as it belongs to Integers.

> It seems to me that this is still open for debate.

That debate has separate context. The equations considered here are different and it are mainly solved with algebraic perspective.

Reply:
What I meant is that I consider this to be problematic:
```
In [1]: solveset(I + Mod(x,5), x, S.Complexes)
Out[1]: {5⋅n - ⅈ | n ∊ ℤ} ∩ ℂ

In [2]: I + Mod(5*0-I, 5) == 0
Out[2]: False
```
`solveset` gives a definite answer, but if I substitute any of the "solutions" they're rejected because in SymPy Mod is still undefined for complex numbers. Given this, I'm not convinced that solveset should accept such equations.

Reply:
Okay, Then I guess for now I will make sure this should return `ConditionSet` as Mod for complex numbers is not defined.